### ISSUE TITLE
Kernel needs more granular endpoints and a memory (statefullness)

### ISSUE THREAD

**Issue by hdoupe** [hdoupe/I-Remote-Procedure-Call#1]
Currently, the kernel endpoints ([`run_nth_year_taxcalc`](https://github.com/open-source-economics/Tax-Calculator/blob/0.20.1/taxcalc/tbi/tbi.py#L98)) are not very "granular" in the sense that you don't have access to many of the functions that are available if you were actually working with the remote package in a non-remote manner. At first, I figured that this would be OK in the short-run and cumbersome in the long-run. However, when thinking through the problem of how to wrap the Tax-Calculator calls in this [script](https://github.com/open-source-economics/OG-USA/blob/0.5.11/ogusa/get_micro_data.py). I realized that this is a serious problem right now.

Of course, this script could be interacted with in the same way that the `run_nth_year_taxcalc` endpoint is interacted with. This would mean the OG-USA package would pass function arguments to the `ogusa_get_micro_data` endpoint in the kernel, which would call a script, `glue_get_micro_data.py`. This would interact with the up-to-date version of Tax-Calculator. You would need to have a separate package containing this script, and you would need to maintain that script and keep it compatible with the Tax-Calculator API and its package requirements, which isn't too different from where we are now.

It would be better if nothing had to be changed on the OG-USA side. This would mean that all of data construction in `get_calculator` and calls to `calc.mtr` could be left as they are now, but instead of making calls directly to the package, they would be making remote calls to the Tax-Calculator package running in the kernel. 

In my mind, this implies the following requirements:
- need a proxy Tax-Calculator package with classes like:
```
class KernelManager():
    ... #manages kernel interactions
    def execute(self, code):
        msg = self.kernel.submit(code)
        return msg
    
    def get(self, obj_name):
         return deserialize(self.kernel.get(obj_name))

class Calculator(KernelManager):
    def __init__(self, **calc arguments):
        self.name = calc_ + uuid()
    
    def mtr(self, *args, **kwargs):
        hash = uuid()
        self.execute("mtr_{hash} = {self.name}.mtr(args, kwargs)") 
        return self.get("mtr_{hash}")
```
- kernel requirements:
  - store previously created objects
  - run arbitrary code
  - return serialized objects upon request

**Comment by hdoupe** [comment]
IMO, the best way to do this is to wrap the iPykernel (and other kernels in the ecosystem). They already do all of this, and they do it well. There's no need to re-invent the wheel. At a first glance, it is pretty easy to interact with the python kernel:

```
from jupyter_client import manager

def run_code(code):
    msg_id = kc.execute(code=code)
    reply = kc.get_shell_msg()
    busy = kc.get_iopub_msg()
    reply['content']
    msg = kc.iopub_channel.get_msg(block=True)
    # print(msg)
    print(msg['content'])
    msg = kc.iopub_channel.get_msg(block=True)
    # print(msg)
    print(msg['content'])

    return msg['content']


if __name__ == '__main__':
    km, kc = manager.start_new_kernel()

    run_code('a = 2')
    run_code("print('hello', a)")

    kc.stop_channels()
    km.shutdown_kernel(now=True)
```

-->

```
HDoupe-MacBook-Pro:I-Remote-Procedure-Call henrydoupe$ python ipykernel_exp.py 
{'code': 'a = 2', 'execution_count': 1}
{'execution_state': 'idle'}
{'code': "print('hello', a)", 'execution_count': 2}
{'name': 'stdout', 'text': 'hello 2\n'}
HDoupe-MacBook-Pro:I-Remote-Procedure-Call henrydoupe$ 
```

The complication here is extracting the data from the iPykernel. My understanding is that all of the data and history is stored in the kernel and only the stdout and stderr are displayed. So, we would have to figure out a way to extract the data from the kernel, serialize it, and deliver it back to the code making the original api call.

**Comment by hdoupe** [comment]
A short-term, super-hacky solution before diving into the iPykernel could be to use something like this as the kernel:
```
LOCAL_VARS = {}

def lowfi_kernel(self, code=None, get_obj=None):
    locals().update(LOCAL_VARS)
    if code is not None:
        eval(code)
    if get_obj is not None:
        obj = locals()[get_obj]

    LOCAL_VARS.update(locals())

    return serialize(obj) 
```

**Comment by hdoupe** [comment]
To illustrate the goal of this, here's a short example of what I think we should be able to do:

```
pol = PolicyProxy()  # creates Policy object in **kernel** named policy_xyz and pol is a PolicyProxy instance
rec = RecordsProxy() # see above
calc = CalculatorProxy.initialize(policy=pol.name, records=rec.name)
```

**Comment by hdoupe** [comment]
I'm going to call #1 resolved for now. You can find the solution at `mesh.proxy`.  Thanks for a very nice solution @lucassz

**Comment by hdoupe** [comment]
Closed #1
